File: 67/python3/main.py
#!/usr/bin/env python3

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        # Initialize the result and carry
        result = []
        carry = 0

        # Pointers for both strings (starting from the end)
        i, j = len(a) - 1, len(b) - 1

        # Iterate as long as there are digits left in either string or carry remains
        while i >= 0 or j >= 0 or carry:
            # Convert the current digit from string to integer, if index is valid
            digit_a = int(a[i]) if i >= 0 else 0
            digit_b = int(b[j]) if j >= 0 else 0

            # Binary addition for current digit and carry
            total = digit_a + digit_b + carry

            # Append the current bit (0 or 1) to the result list
            result.append(str(total % 2))

            # Update carry (it will be 1 if total is 2 or 3)
            carry = total // 2

            # Move to the next digits
            i -= 1
            j -= 1

        # Reverse the result list to get the correct binary string and join it
        return ''.join(reversed(result))

solution = Solution()

# Test case 1: a = "11", b = "1"
result1 = solution.addBinary("11", "1")
print(result1)  # Output: "100"

# Test case 2: a = "1010", b = "1011"
result2 = solution.addBinary("1010", "1011")
print(result2)  # Output: "10101"

# Additional test cases
result3 = solution.addBinary("110", "11")
print(result3)  # Output: "1001"

result4 = solution.addBinary("1111", "1111")
print(result4)  # Output: "11110"

result5 = solution.addBinary("0", "101")
print(result5)  # Output: "101"

File: 67/cpp/main.cpp
#include <iostream>
#include <algorithm>
#include <string>

class Solution {
public:
    std::string addBinary(std::string a, std::string b) {
        std::string result = "";
        int i = a.size() - 1, j = b.size() - 1;
        int carry = 0;

        // Traverse both strings from the end to the beginning
        while (i >= 0 || j >= 0 || carry) {
            int sum = carry;

            // Add bits from a and b, if available
            if (i >= 0) sum += a[i--] - '0';
            if (j >= 0) sum += b[j--] - '0';

            // Compute the current bit and the carry
            result = std::to_string(sum % 2) + result;
            carry = sum / 2;
        }

        return result;
    }
};

int main() {
    Solution sol;

    // Test Case 1
    std::string a1 = "11", b1 = "1";
    std::cout << "Result: " << sol.addBinary(a1, b1) << "\n";  // Expected: "100"
    
    // Test Case 2
    std::string a2 = "1010", b2 = "1011";
    std::cout << "Result: " << sol.addBinary(a2, b2) << "\n";  // Expected: "10101"
    
    // Additional Test Cases
    std::string a3 = "110", b3 = "11";
    std::cout << "Result: " << sol.addBinary(a3, b3) << "\n";  // Expected: "1001"

    std::string a4 = "1111", b4 = "1111";
    std::cout << "Result: " << sol.addBinary(a4, b4) << "\n";  // Expected: "11110"

    std::string a5 = "0", b5 = "101";
    std::cout << "Result: " << sol.addBinary(a5, b5) << "\n";  // Expected: "101"

    return 0;
}
